base/gmp.jl

```diff
@@ -41,11 +41,30 @@ type BigInt <: Integer
     size::Cint
     d::Ptr{Limb}
     function BigInt()
+        if length(bigintpool) > 0
+            b = pop!(bigintpool)
+            finalizer(b, poolingfinalizer)
+            #warn("\n\nREUSING!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!\n\n")
+            #@show b
+            return b
+        end
         b = new(zero(Cint), zero(Cint), C_NULL)
         ccall((:__gmpz_init,:libgmp), Void, (Ptr{BigInt},), &b)
-        finalizer(b, _gmp_clear_func)
+        finalizer(b, poolingfinalizer)
         return b
     end
+    BigInt(alloc, size, d) = new(alloc, size, d)
+end
+
+const bigintpool = BigInt[]
+const BIGINTPOOL_MAXSIZE = 100000
+function poolingfinalizer(b::BigInt)
+    if length(bigintpool) < BIGINTPOOL_MAXSIZE
+        push!(bigintpool, BigInt(b.alloc, b.size, b.d))
```

Comment:
It should work to add `b` itself to the pool. I'm pretty sure the GC can handle that case. At least it used to.